contracts-available/FlightDelayStaking.sol

```diff
@@ -32,230 +48,245 @@ contract FlightDelayRiskPool is Ownable {
-        _result.dip = _s1.dip + _s2.dip;
-        _result.stable = _s1.stable + _s2.stable;
+    function maximumCapacity() public view returns (uint256 _capacity) {
+        return currentStake.div(collatFactor).mul(100);
     }
 
     /**
-     * Calculate the required DIP tokens to stake the requested amount of stable asset
-     * Currently we use a fixed factor e.g. 1.0
-     *
-     * @param _stableStake
-     * @returns _requiredStake Amount of DIP tokens required.
-     *
+     * @dev get total locked stakes
+     * @return _amount is total locked stake
      */
-    function addStake(stake _s1, stake _s2) public pure returns (_stake _result) {
-        _result.dip = _s1.dip + _s2.dip;
-        _result.stable = _s1.stable + _s2.stable;
+    function totalLockedStake() public view returns (uint256 _amount) {
+        return lockedStake;
     }
 
     /**
-     * Calculate the required DIP tokens to stake the requested amount of stable asset
-     * Currently we use a fixed factor e.g. 1.0
-     *
-     * @param _stableStake
-     * @returns _requiredStake Amount of DIP tokens required.
-     *
+     * @dev get available exposure
+     * @return _capacity is available exposure
      */
-    function getStake(address staker)
-        public view
-        returns (stake _stake)
-    {
-        return stakeBalances[staker];
+    function availableCapacity() public view returns (uint256 _capacity) {
+        return maximumCapacity() - lockedStake;
     }
 
     /**
-     * Calculate the required DIP tokens to stake the requested amount of stable asset
-     * Currently we use a fixed factor e.g. 1.0
-     *
-     * @param _stableStake
-     * @returns _requiredStake Amount of DIP tokens required.
+     * @dev unstakes specific amount of staking
+     * @param _stable is the stable amount to unstake
      *
      */
-    function unstake(uint256 _stake, uint256 _stable)
-        public
-        returns (bool _success)
-    {
-        // t.b.d
-        // Unstaking strategy:
-        // Assumption: The items in unstakeRequests are ordered in the sequence they have been recorded.
-        // Pseudocode:
-        //
-        // Check if unstake request is valid, i.e. the contributor has actually staked that many tokens.
-        // WHILE targetStake > lockedStake DO
-        //   Iterate over unstakeRequests until an item is found which can be satisfied.
-        //   IF none is found: RETURN
-        //   Reduce currentStake and targetStake by this item.
-        // DONE
-        //
-        return true;
-    }
+    function unstake(uint256 _stable) public payable {
+        Stake memory _stake = stakeBalances[_msgSender()];
+
+        require(
+            _stake.stable >= _stable,
+            "User staking balance is not enough!"
+        );
 
+        uint256 unlockedStable = getUnlockedStakeFor(_msgSender());
+        uint256 requiredStable = _stable;
 
+        if (requiredStable > unlockedStable) {
```

Comment:
here we need to check if there are already some unstake requests. maybe these can be fulfilled now.